random stuff

franz opened the pull request:
* documentation on driver writing
* clang-tidy config file for C++ conversion
* CHIP-SPV testsuite in examples/
* SPIR-V parser updated to correctly parse Arrays & Struct alignments

pjaaskel commented on doc/sphinx/source/driver_design.rst:
```diff
@@ -0,0 +1,179 @@
+=========================================
+How to write a new device driver for Pocl
```
"PoCL" or "pocl"

pjaaskel commented on doc/sphinx/source/driver_design.rst:
```diff
@@ -0,0 +1,179 @@
+PoCL's internal driver API
+-----------------------------------
+
+Next step is to implement PoCL's internal driver API operations (callbacks) for the device.
```
maybe add (user space) before 'driver' to make it clear we do not have kernel space drivers in PoCL

pjaaskel commented on doc/sphinx/source/driver_design.rst:
```diff
@@ -0,0 +1,179 @@
+
+   these allocate & free device-side memory. See documentation on ``struct _cl_mem->device_ptrs``
+   and ``pocl_mem_identifier``. If the device supports images, these must also allocate/free images.
+   Memory allocation in PoCL is done lazily, IOW a buffer is allocated only on devices will run
```
But still "actively" timing-wise, right? That is, we reserve the space immediately, not delay the allocation to when it's needed (before the first kernel that uses it is launched)? Perhaps clarify that.

pjaaskel commented on doc/sphinx/source/driver_design.rst:
```diff
@@ -0,0 +1,179 @@
+      implementation of ``build_binary``; PoCL will automatically unpack pocl-binaries
+      into the cache directory and setup kernel metadata.
+
+   b. if the device wants to support online (source) compilation using LLVM infrastructure,
```
source -> "OpenCL C"

pjaaskel commented on doc/sphinx/source/driver_design.rst:
```diff
@@ -0,0 +1,179 @@
+      kernel metadata extraction.
+
+   ``build_source`` - required; but only if driver wants to support compilation from source.
+   Note that builds from SPIR-V input are (ATM) handled by the ``build_binary`` callback.
```
ATM -> at the moment

pjaaskel commented on doc/sphinx/source/driver_design.rst:
```diff
@@ -0,0 +1,179 @@
+   Note that builds from SPIR-V input are (ATM) handled by the ``build_binary`` callback.
+
+   ``build_binary`` - required; all drivers (except custom devices) are assumed to support
+   clCreateProgramWithBinary() with at least pocl-binaries as input. If the filed stored
```
filed -> files

pjaaskel commented on doc/sphinx/source/driver_design.rst:
```diff
@@ -0,0 +1,179 @@
+   *strictly* required for a useful driver, since most OpenCL programs don't use these.
+
+   ``supports_binary`` - optional; is called to check whether the driver supports a specific binary
+   note that pocl-binaries are always supported, so this is only called for other binaries.
```
. Note (new sentence)

pjaaskel commented on examples/chip-spv/CMakeLists.txt:
```diff
@@ -0,0 +1,91 @@
+endif()
+
+if(ENABLE_HOST_CPU_DEVICES AND (ENABLE_CONFORMANCE))
+  message(WARNING "CHIP-SPV + CPU driver with ENABLE_CONFORMANCE=ON will fail some tests")
```
Why is this? I'm thinking we should ENABLE_CONFORMANCE=ON by default as soon as we get the (long pending) conformance submission done.

franz replied:
Because enabling conformance disables some features (like image formats) in the CPU driver. We have some buggy/incomplete features, and if we enable them, the CTS tries to test it and it fails. OTOH, CHIP-SPV has tests that need those features.

pjaaskel commented on examples/chip-spv/CMakeLists.txt:
```diff
@@ -0,0 +1,91 @@
+if(CUSTOM_CHIP_GIT_TAG)
+  set(CHIP_GIT_TAG "${CUSTOM_CHIP_GIT_TAG}")
+else()
+  set(CHIP_GIT_TAG "main")
```
Perhaps fix to a commit which we know that should work with the PoCL commit (or we have the known failing cases for). Otherwise regressions in CHIP-SPV cause also the test suite fail here even though it might not be a PoCL issue causing it.

franz replied:
Added

pjaaskel commented on examples/chip-spv/CMakeLists.txt:
```diff
@@ -0,0 +1,91 @@
+add_dependencies(prepare_examples ${TS_NAME})
+
+add_test(NAME "${TS_NAME}"
+         COMMAND "${CMAKE_MAKE_PROGRAM}" check
```
This just runs 'make check' but should it use the list of failing tests in CHIP-SPV as a reference? We could the run it both with PoCL-CPU and PoCL-Level0 at least, if the test can check against the correct failure list.

franz replied:
AFAIK `make check` uses the "full" list of failing tests, by "full" i mean the union of all sets of failing tests.

franz replied:
apparently `make check` is now deprecated, i'll update this to use the new script

franz replied:
fixed

pjaaskel reviewed (changes requested):
At least the known-working CHIP-SPV commit is still to do.

franz commented on lib/CL/devices/spirv_parser.cc:
```diff
@@ -35,32 +36,47 @@
 #define logTrace(...) POCL_MSG_PRINT_INFO(__VA_ARGS__);
 
-typedef std::map<int32_t, std::shared_ptr<OCLFuncInfo>> OCLFuncInfoMap;
+/// Reinterpret the pointed region, starting from BaseAddr +
```
AFAICT `reinterpret_cast` + a load could break if the alignment requirement of the output is larger than that of input, whereas `memcpy` generates unaligned copy instructions which always work. Since this is a template, `memcpy` has to be used.

franz commented:
Rebased on master, also added a few extra fixes: 1) Clang header + linkage check in CMake, 2) extra pointer checks to `clEnqueueSVM` commands, 3) use `scripts/check.py` for CHIP-SPV. Still trying to get a fixed good commit for CHIP-SPV.

franz commented:
rebased on master & added some more fixes. @pjaaskel can you test ?
